### PR TITLE
Feat: pl-order-block ordered grading method now use partial credit

### LINKED ISSUE #6513: Allow pl-order-block to use partial credit for ordered method
In `pl-order-block`, from my understanding, the "ordered" method is similar to the ranking/dag methods where the ranking is implicitly the order of correct answers. With that in mind, would it make sense to include partial credit support for the ordered method using the same principle as ranking or dag?

There is, of course, the workaround, which is to set the method to ranking and give each answer a strictly increasing ranking number, but since these are equivalent it might be simpler to just allow the same partial credit to be used.

### PR DESCRIPTION
**Pull request by Frozemint**
### **This PR supersedes #8151, because 8151 has been stale for a long time and something funny happened to its commit history**.

Per #6513, the ordered grading method for pl-order-block currently uses an "all-or-nothing" approach for grading answers. Either the student gets 100% for correctly ordering all of the answer, or the student gets 0%.

This PR implements partial credit support to the ordered grading method by using the workaround mentioned in the original issue: the grading code now add a strictly increasing `ranking` number to the `true_answer_list` when the grading method is set to ordered, and then reuse the grading logic for `ranking` and `dag` to grade the ordered submission. This PR will close #6513.

For backward compatibility as specified in the previous discussions here in the superseded PR #8001, partial credits is disabled for ordered grading by default, because partial credits for ordered did not exist before. Partial credit is only enabled when the user specify both `grading-mode='ordered'` and `partial-credit='lcs'`.

The documentation in `docs/elements.md` is also updated according to the new feature implemented in this PR.

### CODE REVIEW

File: docs/elements.md [hunk 1]
```diff
@@ -879,7 +879,7 @@ Element to arrange given blocks of code or text that are displayed initially in
...
+| `partial-credit`      | "none" or "lcs"                                 | Depends on `grading-method`    | For the `dag`, `ordered`, and `ranking` grading methods, you may specify `none` for no partial credit or `lcs` (default) for partial credit based on the LCS edit-distance from the student solution to some correct solution. Note that `partial-credit` defaults to `lcs` for the `dag` and `ranking` grading method, and to `none` for the `ordered` grading method for backward compatibility. For the other grading methods, using this property is not yet supported. Grading method `unordered` will always assign partial credit, and grading method `ordered` will never do so.                                                       |
```
Comment (SethPoulsen) [inline docs/elements.md:882]: you should get rid of the '(default)' in parentheses after the first mention of  'lcs'. You also need to remove the last clause that says 'grading method 'ordered' will never do so'

File: testCourse/questions/orderBlocks/question.html [hunk 2]
```diff
@@ -98,3 +98,20 @@
...
+</pl-order-blocks>
+
+<pl-order-blocks
+answers-name="test-ordered-grading-method-partial-credit"
+grading-method="ordered"
+partial-credit="lcs">
+<pl-answer correct="true">first</pl-answer>
+<pl-answer correct="true">second</pl-answer>
+<pl-answer correct="true">third</pl-answer>
+</pl-order-blocks>
```
Comment (SethPoulsen) [inline testCourse/questions/orderBlocks/question.html:117]: can you keep the formatting/indentation here consistent with the rest of the file?

File: apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py [hunk 3]
```diff
@@ -72,6 +72,9 @@ class OrderBlocksAnswerData(TypedDict):
 FIRST_WRONG_TYPES = frozenset(
     [FeedbackType.FIRST_WRONG, FeedbackType.FIRST_WRONG_VERBOSE]
 )
+RANKING_DAG_ORDERED_GRADING_TYPE = frozenset(
+    [GradingMethodType.RANKING, GradingMethodType.DAG, GradingMethodType.ORDERED]
+)
```
Comment (SethPoulsen) [inline apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:77]: I don't like the name of this array, can we use something more descriptive, like 'LCS_GRADABLE_TYPES' or something?

File: docs/elements.md (hunk 1 above)
Comment (Frozemint) [reply docs/elements.md]: Fixed in [6a4316d](https://github.com/PrairieLearn/PrairieLearn/pull/9292/commits/6a4316d898b52a3bc7ba5a707bce4107268498e8).

File: testCourse/questions/orderBlocks/question.html (hunk 2 above)
Comment (Frozemint) [reply testCourse/questions/orderBlocks/question.html]: fixed in [24dbf38](https://github.com/PrairieLearn/PrairieLearn/pull/9292/commits/24dbf38c5226346e5fb55ab48562a10754a69d4a).

File: apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py [hunk 4]
```diff
@@ -902,6 +907,18 @@ def grade(element_html: str, data: pl.QuestionData) -> None:
...
+    # For backward compatibility, we need to override the default partial credit type
+    # when grading_method = ORDERED
+    return (
+        PartialCreditType.NONE
+        if grading_method is GradingMethodType.ORDERED
+        else PartialCreditType.LCS
+    )
+
+
 def test(element_html: str, data: pl.ElementTestData) -> None:
```
Comment (SethPoulsen) [inline apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:922]: Currently these automated tests are only testing a couple of hard coded cases. Did you try submitting a bunch of different answers to a few different types of 'ordered' questions just to make sure everything is working right?

I know this is not ideal, as you can see from the TODOs here, I would prefer to have some randomized fuzz testing here so we don't have to do this manually.

File: apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py (hunk 3 above)
Comment (Frozemint) [reply apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py]: fixed in [955ce68](https://github.com/PrairieLearn/PrairieLearn/pull/9292/commits/955ce68fe1b5f068d86c76cf1eae3e1101918dfd).

File: apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py (hunk 4 above)
Comment (SethPoulsen) [reply apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:922]: excellent thanks!

File: apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py [hunk 5]
```diff
@@ -72,6 +72,9 @@ class OrderBlocksAnswerData(TypedDict):
 FIRST_WRONG_TYPES = frozenset(
     [FeedbackType.FIRST_WRONG, FeedbackType.FIRST_WRONG_VERBOSE]
 )
+LCS_GRADABLE_TYPE = frozenset(
```
Comment (SethPoulsen) [inline apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:75]: should be 'TYPES' like the one above it, not 'TYPE'

File: docs/elements.md [hunk 6]
```diff
@@ -865,25 +865,25 @@ Element to arrange given blocks of code or text that are displayed initially in
...
+| `partial-credit`      | "none" or "lcs"                                 | Depends on `grading-method`    | For the `dag`, `ordered`, and `ranking` grading methods, you may specify `none` for no partial credit or `lcs` for partial credit based on the LCS edit-distance from the student solution to some correct solution. Note that `partial-credit` defaults to `lcs` for the `dag` and `ranking` grading method, and to `none` for the `ordered` grading method for backward compatibility. For the other grading methods, using this property is not yet supported. Grading method `unordered` will always assign partial credit. |
```
Comment (SethPoulsen) [inline docs/elements.md:882]: you should just remove the sentence 'For the other grading methods, using this property is not yet supported.' Its a bit awkward sounding, now that there is only one 'other' grading method, and we already make it clear what happens with that one in the next sentence.

File: apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py (hunk 5 above)
Comment (Frozemint) [reply apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py]: Fixed [2e87c5b](https://github.com/PrairieLearn/PrairieLearn/pull/9292/commits/2e87c5b944528607baf0bcc04c1665f93bc9315d)

**Comment by SethPoulsen** [comment]
Just a couple more minor things and I will approve. Thanks for your work on this @Frozemint !

File: docs/elements.md (hunk 6 above)
Comment (Frozemint) [reply docs/elements.md]: Fixed in [04ad4e5](https://github.com/PrairieLearn/PrairieLearn/pull/9292/commits/04ad4e56c0b854239b8f727954053ead55b70cc0).

**Comment by Frozemint** [comment]
Of course, this is quite overdue. Glad to be able to help again now that school is settling down.

**Review by SethPoulsen — APPROVED** [review]

**Comment by SethPoulsen** [comment]
Looks like the javascript linter is failling on the documentation file. Once you fix (by running prettier on it), I can merge.

**Pull request merged**